### ISSUE TITLE
Failed to write revision with attachment

### ISSUE THREAD

**Issue by ryuzmukhametov** [couchbase/couchbase-lite-ios#802]
In my app sometimes when a document with attachment is pulled I get next warning:
> 19:06:15.987‖ WARNING: CBL_Puller[http://localhost:5984/a] failed to write {my_doc_id #4-a0716cf34b240d1dec1ef7ada408063e}: status=491

As result the pull replication is stopped. Moreover it is impossible to start pull replication even it use start or restart methods.

Next steps show how reproduce it in 'CBL iOS Test App' (sha commit: 9aa3bce43341d42df014595fdfbdb3bdc9f29e61)
- Configure the app with syncpoint
- Launch the application and create a document with attachment (see 'createDocument With Attachment' below). The document is pushed to the server.
- Stop application
- Change document many times in the server (more than one time)
- Start the app

When the app start it can't pull revision from remote server.

Sample code for create document with attachment (I call it from viewWillAppear):

``` Objective-C
- (void)createDocumentWithAttachment
{
    NSString *docId = @"my_doc_id";
    CBLDatabase *db = self.database;
    CBLDocument *doc = [db documentWithID:docId];
    CBLSavedRevision *rev = [doc putProperties:@{@"key":@"value"} error:nil];
    NSLog(@"rev = %@", rev.revisionID);
    NSString* str = @"teststring";
    NSData* data = [str dataUsingEncoding:NSUTF8StringEncoding];
    NSString *contentType = @"text/plain";
    CBLSavedRevision *currentRevision = rev;
    CBLUnsavedRevision *unsavedRevision = [currentRevision createRevision];
    [unsavedRevision setAttachmentNamed:@"my_attachment" withContentType:contentType content:data];
    currentRevision = [unsavedRevision save:nil];
    NSLog(@"%@", currentRevision.revisionID);
}
```
It occurs when I use SQLite and ForestDB data storage.

**Comment by ryuzmukhametov** [comment]
Here is a second time launch application log (use CBLDatabase, ChangeTracker, RemoteRequest, Sync log chanels) https://gist.github.com/ryuzmukhametov/cad86667b86d993348be

**Comment by snej** [comment]
491 is `kCBLStatusBadAttachment` so something's going wrong saving the attachment.

What is the exact server you're using?

**Comment by snej** [comment]
Thanks for the logs. From the server URL `http://localhost:5984` I'm guessing you are replicating with CouchDB. (What version?)

This sounds like #786 except you're not getting the same warning message. But it's probably also related to the fact that recent versions of CouchDB have started including MIME headers in their attachment multipart bodies.

**Comment by ryuzmukhametov** [comment]
I use Apache CouchDB version 1.6.1

**Comment by snej** [comment]
OK, I can reproduce it (with the latest master branch). The error occurs at CBLDatabase+Attachments.m:350:
```objc
                    if (status == kCBLStatusOK || status == kCBLStatusNotFound)
                        status = kCBLStatusBadAttachment;
```
The attachment being processed is a stub but the local db doesn't have the parent revision, so it fails. It should be looking farther back in the rev history, since in this case there are several missing revs between the one being inserted and the last one that's known locally.

**Comment by snej** [comment]
That was a nasty little edge case — thanks for reporting it. Only happened when pulling from CouchDB, because CouchDB identifies attachments with MD5 digests instead of SHA-1, so our code can't look up the attachment by its digest.

**Comment by snej** [comment]
FYI, you should be able to cherry-pick this fix onto the 1.1 branch. At this point that would be safer than trying to use the master branch (too many brand-new changes.) You'll also want to pick up the other CouchDB compatibility fix (see #786.)

**Comment by ryuzmukhametov** [comment]
Thank you, Jens, now it works for me!
